src/violation.rs

```diff
@@ -1,15 +1,98 @@
-// TODO: Implement Violations for constraints
-#[derive(Debug, Clone)]
-pub struct Violations {
-    violations: Vec<Violation>,
+/// Represents Violations found during validation of values based on schema types
+pub trait Violations {
```

Comment:
Sure. I can rename these.

Reply:
> question—Is it possible to make these into immutable, plain-old data objects?

I second this question. Making this a trait suggests that we intend to have several implementations of the `Violations` collections, but I'm not sure I understand why we'd do that. I expected `Violations` to be a simple wrapper around `Vec<Violation>`. Having this be a trait requires the `validate` method to be generic by accepting `&mut impl Violations` as a parameter.

Reply:
Because the types can have constraints which can have type references, removing `trait` `Violations` would mean, I will not be able to pass `ViolationsImpl`(`ViolationTree`: which has no message) and `ViolationImpl`(`ViolationLeaf`: which has more details on where exactly the issue was found through message) interchangeably. The calls to validate method of constraints/types should expect the same struct/trait hence I have Violations as a trait. Is there any other way that you suggest to use Violations here? @zslayton @popematt.

Reply:
> Having this be a trait requires the validate method to be generic by accepting &mut impl Violations as a parameter.

If `validate` is a pure function (as I suggest elsewhere), then it shouldn't have to accept `Violations` as a parameter. Types and Constraints should be able to have a signature something like this:

```rust
fn validate(&self, value: &OwnedElement) -> Option<Violation>
```

> I will not be able to pass ViolationsImpl(ViolationTree: which has no message) and ViolationImpl(ViolationLeaf: which has more details on where exactly the issue was found through message)

I don't think it's worth keeping around any Violations implementation that has no message. We can simplify violations by having only one type of `Violation` that can have 0..many children and can be used for the root, interior, and leaf nodes of the tree.

Reply:
I understand it now. Yes. In that case I will remove `trait Violations` and just keep single `struct Violation` which can have a message as well as `Vec<Violation>` (describing children violations)